UPGRADE_GUIDE.md

```diff
@@ -2,8 +2,133 @@
 
 Use the following guide to assist in the upgrade process of the `easypost-ruby` library between major versions.
 
+- [Upgrading from 4.x to 5.0](#upgrading-from-4x-to-50)
+- [Upgrading from 3.x to 4.0](#upgrading-from-3x-to-40)
+
+## Upgrading from 4.x to 5.0
+
+### 5.0 High Impact Changes
+
+- [Thread-Safe With Client Object](#50-thread-safe-with-client-object)
+- [Updated Dependencies](#50-updated-dependencies)
+- [Improved Error Handling](#50-improved-error-handling)
+
+### 5.0 Medium Impact Changes
+
+- [Corrected Naming Conventions](#50-corrected-naming-conventions)
+
+### 5.0 Low Impact Changes
+
+- [Beta Namespace Changed](#50-beta-namespace-changed)
+
+## 5.0 Thread-Safe with Client Object
+
+Likelihood of Impact: High
+
+This library is now thread-safe with the introduction of a new `Client` object. Instead of defining a global API key that all requests use, you now create an `Client` object and pass your API key to it with optional open and read timeout params. You then call your desired functions against a `service` which coincide with the `Client` object:
+
+```ruby
+# Old way
+require 'easypost'
+
+EasyPost.api_key = ENV['EASYPOST_API_KEY']
+shipment = EasyPost::Shipment.retrieve('shp_...')
+
+# New way
+require 'easypost'
+
+client = EasyPost::Client.new(api_key: ENV['EASYPOST_TEST_API_KEY'])
+shipment = client.shipment.retrieve('shp_...')
```

Comment:
suggestion: We want to keep it simple and just talk about invocation here probably

```suggestion
# Old way
EasyPost.api_key = ENV['EASYPOST_API_KEY']
shipment = EasyPost::Shipment.retrieve('shp_...')

# New way
client = EasyPost::Client.new(api_key: ENV['EASYPOST_TEST_API_KEY'])
shipment = client.shipment.retrieve('shp_...')
```